Changes for OWLS-83431

ankedia opened the pull request:
Changes for OWLS-83431 - Create Pods one-at-a-time waiting on scheduling

russgold commented:
Can you explain the logic in ManagedServerUpIteratorStep.apply() ?

ankedia commented:
> Can you explain the logic in ManagedServerUpIteratorStep.apply() ?

ankedia commented:
> Can you explain the logic in ManagedServerUpIteratorStep.apply() ?

The apply method runs StartManagedServersStep (which starts managed servers) for 2 different types of servers -
1. Non-clustered (standalone) servers
2. Clustered servers from one or more clusters

It creates a child fiber for each cluster and a fiber for standalone (non-clustered) servers and it executes StartManagedServersStep. Since each cluster can have a different value for "maxConcurrentStartups", I'm creating different fiber for each cluster (and for non-clustered servers). The idea for maxConcurrent startup is that only n number of pods can be created simultaneously until all previous pods are ready.

In StartManagedServersStep, I check to see how many pods are scheduled and how many pods are ready. This is used to decide if a new managed server pod can be created. Each managed server pod is created in a separate child fiber. Please let me know if this is not clear or if you have any questions. Thanks.

russgold commented:
It's clear but incomplete. I'm trying to understand the logic. Can you describe it in pseudocode? What is the logic in the method?

ankedia commented:
> It's clear but incomplete. I'm trying to understand the logic. Can you describe it in pseudocode? What is the logic in the method?



> It's clear but incomplete. I'm trying to understand the logic. Can you describe it in pseudocode? What is the logic in the method?
ok, I will work on that. Do you or @alai8 know if there's any psuedocode or document for original ManagedServerUpIteratorStep or other related steps? It'll help me with the terminology and structure. Thanks.

ankedia commented:
> It's clear but incomplete. I'm trying to understand the logic. Can you describe it in pseudocode? What is the logic in the method?
ok, I will work on that. Do you or @alai8 know if there's any psuedocode or document for original ManagedServerUpIteratorStep or other related steps? It'll help me with the terminology and structure. Thanks.

russgold commented:
There is no formalism for pseudocode. I'm just asking you to describe what is going on in english, not at the code level.

Let me get specific. The body of the method is very roughly
1. The first condition is whether the following step for the head of the queue is a ServerDownStep. Why? What does that mean? It seems a very brittle thing to test
2. If the queue is empty you advance to the after-step
3. The next condition (after verifying that the queue is not empty) uses some combination of the number of scheduled pods and the ready pods to decide to fork join. Why? What is the intent?
4. The third condition decides to delay if the queue is non-empty. Why delay? How do you know that there is nothing left to process? Is there any way for it to be empty at this point, given that you already checked that (see 2)
5. Once again, if the queue is empty you advance to the after-step, which seems redundant

What have I missed?

ankedia commented:
Step StartManagedServerStep (Argument clusterName, Argument Collection<StepAndPacket> startDetails, Argument nextStep){ 

    Initialize current scheduled count for cluster "clusterName" to 0 
	Initialize FIFO Queue - it'll hold StepsAndPackets for starting managed servers for this cluster
    
    Constructor {
	  Populate FIFO queue containing StepsAndPackets from startDetails to satrt servers in cluster "clusterName".
	  Get maxConcurrency for this cluster
    }    
	
	apply method {
	  Check first StepAndPacket in queue
	  if (step in StepsAndPackets is for bringing server down i.e. for scale down scenario) {
	    start a fiber to run ServerDown step and then invoke this step again
	  } else if (all steps in startDetails completed) {
	    proceed to nextStep
      } 	  
	  
	  // This makes assumption that admin server is already scheduled 
	  if (current scheduled count is less than total scheduled servers for cluster "clusterName") { //initial value of total scheduled servers is 1
	    if (current scheduled count is less than (maxConcurrency + ready managed servers for cluster "clusterName")) {
		   Dequeue StepAndPacket from FIFO queue
		   Increment current scheduled count
		   Start a child fiber to run step for managed server pod and service creation and then invoke this step again
		}
	  }
	  if (More items in queue) {
	     Wait 100 mill-seconds to give a chance for previous server to get scheduled and then invoke this step again
      } else { // All managed server start steps executed
	     proceed to nextstep
	  }
	} 
}

ankedia commented:
1. The `bringManagedServerUp()` has below existing code. My understanding of this code is that it checks ServerStartupInfo to see if only service needs be created for managed server (i.e. when pre-creating server service) then it creates a ServerDownStep in startDetails. This is also executed during scale down scenario to bring managed server down based on ServerStartupInfo. Previously it was executing all steps in startDetails array list in parallel. I added check make sure it executes ServerDownStep before we check for scheduled and ready servers. 

```
  private static Step bringManagedServerUp(ServerStartupInfo ssi) {
    return ssi.isServiceOnly()
        ? ServiceHelper.createForServerStep(
            true, new ServerDownStep(ssi.getServerName(), true, null))
        : ServiceHelper.createForServerStep(PodHelper.createManagedPodStep(null));
  }
```
The queue empty check in "else" condition is to handle cases where startDetails only contains ServerDownSteps and no steps to create managed server pods/service. In this case, we don't need to rely on scheduled count etc. I can look to change/refactor this code if it's brittle to test.

3. Please let me know if intent is not clear from above pseudocode.

4. This is executed after checking the condition of previous pods to see if they're scheduled and/or ready. If previous pod is not yet scheduled, it'll wait 100ms and invoke this step again

5. Again, this check is after checking the condition of previous pods. I had added it before I added the check for ServerDownStep and previous check to see if queue is empty. I forgot to revisit this after adding check for ServerDownStep, I'll check and remove it.

russgold commented:
The first simplification I see is that you should check if the queue is empty first. If it is not, there is no need to check if peek returns a null, as it won't.

I would then extract the check on that first item into a method with a name such as isShuttingDown. It will make it clear as to what is going on much better than reading the computation logic. And since that logic does a lot of null checks, Optional.ofNullable would probably make it easier to read the implementation.

The next pair of nested ifs seems to be the heart of the functionality. My first observation is that it should really be a single if with a complex condition extracted into a method for clarity. I suppose it means something like "there are more servers to start in the cluster"? It is important to make it as easy to understand as possible, and right now it isn't obvious to me why this logic is the one to use. What does the scheduled count mean? Isn't it actually the number of pods in the cluster that have been started, not scheduled? Also, I suspect this could be another else if.

ankedia commented:
Thanks Russ for the comments and suggestions. I have made changes to incorporate these suggestions. Please let me know if I missed something.

russgold commented:
It's definitely improved. I would further suggest:
1. Use isEmpty() rather than size() = 0
2. Combine the second group to the first using if else
3. You don't need the local variable servers.

So, for example:

`    @Override
    public NextAction apply(Packet packet) {
      if (startDetailsQueue.isEmpty()) {
        return doNext(new ManagedServerUpAfterStep(getNext()), packet);
      } else if (isServiceOnlyOrShuttingDown()) {
        return doForkJoin(this, packet, Collections.singletonList(startDetailsQueue.poll()));
      } else if (serverAvailableToStart(packet.getSpi(DomainPresenceInfo.class))) {
        this.countStarted++;
        return doForkJoin(this, packet, Collections.singletonList(startDetailsQueue.poll()));
      } else {
        return doDelay(this, packet, 100, TimeUnit.MILLISECONDS);
      }
    }
`

Why is `maxConcurrency` a variable? What values can it take?

russgold commented:
A suggest about naming variables: Use a consistent pattern such as "numXXX." Right now you have `countStarted` and `readyManagedPods`

ankedia commented:
I have made changes as per above suggestions. `maxConcurrency ` is the value of `MaxConcurrentStartup` cluster configuration in domain. It's an instance variable and I have changed canStartConcurrently() to refer it by this. Thanks.

rjeberhard commented on operator/src/main/java/oracle/kubernetes/operator/helpers/PodHelper.java:
```diff
@@ -85,6 +86,59 @@ public static boolean isReady(V1Pod pod) {
+  /**
+   * Get list of scheduled pods.
+   * @param info Domain presence info
+   * @return list containing scheduled pods
```
Missing Javadoc @param for clusterName

ankedia replied:
Fixed.

rjeberhard commented on operator/src/main/java/oracle/kubernetes/operator/helpers/PodHelper.java:
```diff
@@ -85,6 +86,59 @@ public static boolean isReady(V1Pod pod) {
+   * @param info Domain presence info
+   * @return list containing scheduled pods
+   */
+  public static List<String> getScheduledPods(DomainPresenceInfo info, String clusterName) {
```
I'll verify below, but it's surprising that you have a method to list scheduled pods. I would think that you would be interested in unscheduled pods.

rjeberhard commented on operator/src/main/java/oracle/kubernetes/operator/helpers/PodHelper.java:
```diff
@@ -85,6 +86,59 @@ public static boolean isReady(V1Pod pod) {
+   * @return true, if pod is scheduled
+   */
+  public static boolean getScheduledStatus(V1Pod pod) {
+    V1PodSpec status = pod.getSpec();
```
Fix variable name

ankedia replied:
Fixed.

rjeberhard commented on operator/src/main/java/oracle/kubernetes/operator/steps/ManagedServerUpIteratorStep.java:
```diff
@@ -142,19 +154,53 @@ private boolean isServerInCluster(ServerStartupInfo ssi) {
+    }
+
+    private boolean serverAvailableToStart(DomainPresenceInfo info) {
+      return ((this.numStarted < PodHelper.getScheduledPods(info, clusterName).size())
```
What do you think about switching this to test PodHelper.unscheduledPodsExist() rather than comparing the count of scheduled pods to an expected number?

ankedia replied:
I believe the reason I went with current `count` based approach is because we can only schedule/start `n` number pods in parallel based on value of maxConcurrency. Previously the code was creating `maxConcurrency` number of fibers and fibers used to wait until pod becomes ready. With current approach, I don't make the fibers wait until pod is ready but start servers (create fibers) only if all previous pods are scheduled and it matches the criteria for `maxConcurrency`. I think it's possible to test  PodHelper.unscheduledPodsExist() but we need to check for number of ready pods before deciding to start a new server (start a new fiber). I can try the approach you suggested and test it out for different use-case scenarios.

rjeberhard commented on operator/src/main/java/oracle/kubernetes/operator/steps/ManagedServerUpIteratorStep.java:
```diff
@@ -142,19 +154,53 @@ private boolean isServerInCluster(ServerStartupInfo ssi) {
+        Collection<StepAndPacket> servers = Collections.singletonList(startDetailsQueue.poll());
+        return doForkJoin(this, packet, servers);
+      } else if (serverAvailableToStart(packet.getSpi(DomainPresenceInfo.class))) {
+        this.numStarted++;
```
What is the concurrency of this code? Will multiple threads ever be running here?

ankedia replied:
It's one thread per cluster and one thread for non-clustered servers. It'll be single threaded for one cluster and 2 threads for 2 cluster case. If domain also have non-clustered (standalone) servers, it'll create a separate fiber for those servers.

rjeberhard replied:
Let me be clearer in my question. Do I need to worry about the concurrency of updating the numStarted variable? I can't really tell from your answer.

ankedia replied:
Sorry for not being clear in my previous answer. In 2 cluster scenario, 2 threads could reading/updating in parallel but overall concurrency will be low. I have `numStarted` to AtomicInteger as we discussed and made `numConcurrency ` final by setting it in constructor. Please let me know if I missed anything. Thanks.

rjeberhard commented on operator/src/main/java/oracle/kubernetes/operator/steps/ManagedServerUpIteratorStep.java:
```diff
@@ -142,19 +154,53 @@ private boolean isServerInCluster(ServerStartupInfo ssi) {
+    }
+
+    private boolean canStartConcurrently(int numReady) {
+      return ((this.maxConcurrency > 0) && (this.numStarted < (this.maxConcurrency + numReady - 1)))
```
Similarly, it seems like you ought to be checking for the number of unready pods.

ankedia replied:
Please see above comment about creating fibers based on value of `maxConcurrency`. I think we can't wait for all previous pods to be ready and need to check for number of ready pods in order to decide whether to create a new fiber/thread or not.

ankedia commented:
Thanks @rjeberhard for your comments and suggestions. I have addressed some of your comments and have replied to other comments. I have thoroughly tested current approach by running functional, integration and unit tests. If we decide to change the approach, these tests will need to be rerun. Please let me know your thoughts.

rjeberhard commented:
Okay, I'm happy enough with the current approach.
